x/ecocredit/server/genesis.go

```diff
@@ -1,31 +1,189 @@
 package server
 
 import (
+	"bytes"
+	"encoding/binary"
 	"encoding/json"
 
 	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/pkg/errors"
+	"github.com/regen-network/regen-ledger/orm"
+	"github.com/regen-network/regen-ledger/types/math"
 	abci "github.com/tendermint/tendermint/abci/types"
 
 	"github.com/regen-network/regen-ledger/types"
 	"github.com/regen-network/regen-ledger/x/ecocredit"
 )
 
+type supplyKey func(batchDenom batchDenomT) []byte
+type balanceKey func(addr sdk.AccAddress, batchDenom batchDenomT) []byte
+
 // InitGenesis performs genesis initialization for the ecocredit module. It
 // returns no validator updates.
 func (s serverImpl) InitGenesis(ctx types.Context, cdc codec.Codec, data json.RawMessage) ([]abci.ValidatorUpdate, error) {
 	var genesisState ecocredit.GenesisState
 	cdc.MustUnmarshalJSON(data, &genesisState)
+
+	s.idSeq.InitVal(ctx, genesisState.IdSeq)
+
 	s.paramSpace.SetParamSet(ctx.Context, &genesisState.Params)
+
+	if err := orm.ImportTableData(ctx, s.classInfoTable, genesisState.ClassInfo, 0); err != nil {
+		return nil, errors.Wrap(err, "class-info")
+	}
+
+	if err := orm.ImportTableData(ctx, s.batchInfoTable, genesisState.BatchInfo, 0); err != nil {
+		return nil, errors.Wrap(err, "batch-info")
+	}
+
+	store := ctx.KVStore(s.storeKey)
+
+	if err := setBalance(store, genesisState.TradableBalances, func(addr sdk.AccAddress, batchDenom batchDenomT) []byte {
+		return TradableBalanceKey(addr, batchDenom)
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := setBalance(store, genesisState.RetiredBalances, func(addr sdk.AccAddress, batchDenom batchDenomT) []byte {
+		return RetiredBalanceKey(addr, batchDenom)
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := setSupply(store, genesisState.TradableSupplies, func(bd batchDenomT) []byte {
+		return TradableSupplyKey(bd)
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := setSupply(store, genesisState.RetiredSupplies, func(bd batchDenomT) []byte {
+		return RetiredSupplyKey(bd)
+	}); err != nil {
+		return nil, err
+	}
```

Comment:
Supplies should only be used as an invariant check. Setting balances should update the supply and then we should check that the supply is what is expected. We should never just set supply based on a value in genesis.